### DISCUSSION TITLE
Docgen (Murex document generator)

### DISCUSSION THREAD

**Discussion by lmorg** [lmorg/murex#740] [Ideas]
Docgen is a tool I'd written to speed up writing Murex documentation. It's kind of like a CMS where each page is written in YAML and that gets merged, kind of like a mail merge of old, into templates to generate markdown files. It's also been used to autogenerate source code too however a lot of those use have disappeared more recently.

Docgen does have its rough edges but I do still think templating the docs is generally a good idea because it leads to consistency in the documentation. And while there are definite pain points in docgen, it's primary purpose was to enabled me to write a lot of documentation quickly. At least for that purpose, it succeeded extremely well.

That all said, I do think we are at a good point to review the quirks of docgen and see what I can do to improve the situation. So I'm opening a discussion on this here.

**Comment by lmorg** [comment]
One initial thought is that i could make IDs unique. Currently an ID is only unique to a category, which makes moving pages between categories very painful. Having a globally unique ID would simplify that

**Reply by orefalo** [reply]
Agreed,
the id could be any string, in the current doc, I would opt for `path/id`

another important feature to add - linked content

you have synonyms such as `![` which are equivalent to index.
If I want to add them to the left side panel today, I can't point to the same content as index.

**Reply by orefalo** [reply]
I must also point that I am in the process of adding a `<path>/id` to all references sections
A major task...

but it helps the moving

**Reply by lmorg** [reply]
`<path>/id` is how it's already working. Are you able to describe a little more what your thought is here and the changes you're making?

**Reply by orefalo** [reply]
I am converting this,

```
  Related:
  - prepend
  - append
  - map
  - a
  - ja
  - jsplit
  - count
  - item-index
  - range
  - mtac
  - msort
  - types/json
```
into that
```
  Related:
  - commands/prepend
  - commands/append
  - commands/map
  - commands/a
  - commands/ja
  - commands/jsplit
  - commands/count
  - commands/item-index
  - parser/range
  - commands/mtac
  - commands/msort
  - types/json
```

to make page movements easier,
  with the latest, a simple search/replace is a catch all.

**Reply by orefalo** [reply]
What you should do moving forward, is removing the category, or rather enforcing it as the first element of the id pattern.

from
```
 DocumentID: 2darray
  Title: >-
    `2darray` 
  CategoryID: commands
  Summary: >-
```
to
```
 DocumentID: commands/2darray
  Title: >-
    `2darray` 
  Summary: >-
```

**Reply by lmorg** [reply]
`docgen -warning` will report on any documents that couldn’t be located. 

Personally I’d have gone the other way and stripped the category ID from the `related` fields. It makes more sense the document ID being unique across all categories rather than just its own category. That way you can change the category ID of a document and everything else updates automatically without you having to make additional adjustments.

**Reply by orefalo** [reply]
I am missing one piece of the puzzle to align with you.
The end state is a proper classification (in vuepress), where the pages land in their respective category.

If the category is stripped - either as a field, or from the id... how will it work?

Update: think I am starting to understand what you mean

```
 DocumentID: 2darray
  Title: >-
    `2darray` 
  CategoryID: commands
  Summary: >-
```

what you are saying is that any ref to 2darray implicitly references the category

Indeed, it will ease the major reediting I am performing now.

**Reply by orefalo** [reply]
I just did the changes, of course it won't generate until the docgen code is modified.

**Reply by lmorg** [reply]
Awesome, thank you. It will probably be a couple of days before I get to docgen because i have a few personal commitments coming up. But I'll definitely get this merged asap.

**Reply by lmorg** [reply]
I've made the update to docgen and merged your changes however there are a few documents sharing the same documentID. So I'll go through and fix that before merging everything back into `develop`

**Reply by lmorg** [reply]
Everything should be merged into `develop` and so changes in https://dev.murex.rocks as well (albeit I haven't had a chance to check to confirm)

Sorry it's taken me a while to get to this. It's been really busy at work and home this last week or so.

**Reply by orefalo** [reply]
ah thank you - will pick up later this week. playing catch up

**Reply by lmorg** [reply]
absolutely no rush. I think I'm going to slow my pace of development on Murex for the next couple of months anyways (lots going on outside of work). Plus v5.0 feels like a pretty stable release so hopefully that has bought me a little bit of time to strategise :)
